contributed/charlieplexer/main.js

```diff
@@ -0,0 +1,31 @@
+class Charlieplexer {
+  constructor(pins) {
+    this.initializePins(pins);
+  }
+
+  initializePins(pins) {
+    this.pins = pins.map(p => new Digital(p, Digital.Input));
+  }
+
+  scan() {
+    for (let i = 0; i < this.pins.length - 1; i++) {
+      for (let j = i + 1; j < this.pins.length; j++) {
+        const p1 = this.pins[i];
+        const p2 = this.pins[j];
```

Comment:
FWIW - if you expect a large number of pins, we would use a local for the length. XS doesn't have the optimizations of V8 and friends, so minimizing property look-ups helps.

```
   const pins = this.pins;
    for (let i = 0, length = pins.length; i < length - 1; i++) {
        for (let j = i + 1; j < length; j++) {
        const p1 = pins[i];
        const p2 = pins[j];
```

Here it the number is generally small, so no problem.

Reply:
@phoddie this is a wonderful point - I'm so used to v8-alike's optimizations it affects the way I code. Is there a document that goes over advice like this? I imagine there's a number of anti-patterns once you know which optimizations are present or not in xs.

Reply:
No document at the moment. It would make a good blog post. I've added it to the list on the wall!

A good rule is to think of XS more like a non-optimizing compiler: the code you write is close to what will execute. There are some optimizations, of course, but surely no runtime execution monitoring to optimize the JIT generated code. ;)